src/solvers/elliptic/linearSolver/CHEBYSHEV.cpp

```diff
@@ -0,0 +1,204 @@
+    const dfloat rCoeff = 2.0 * rho / delta;
+    const dfloat dCoeff = rho * rho_prev;
+    // d = rho*rho_prev * d + 2 * rho / delta * r
+    platform->linAlg->axpbyMany(Nlocal, Nfields, offset, rCoeff, o_r, dCoeff, o_d);    
+    rho_prev = rho;
+  }
+
+  // x = x + d
+  platform->linAlg->axpbyMany(Nlocal, Nfields, offset, one, o_d, one, o_z);
+
+  if(!options.compareArgs("PRECONDITIONER", "NONE")) {
+    ellipticPreconditioner(elliptic, o_z, o_x);
+  }
+}
+
+int chebyshev_aux(elliptic_t* elliptic, occa::memory &o_r, occa::memory &o_x,
+        const dfloat tol, const int MAXIT, dfloat &rdotr, const int tstep)
+{
+  mesh_t* mesh = elliptic->mesh;
+  setupAide& options = elliptic->options;
+  const int verbose = platform->options.compareArgs("VERBOSE", "TRUE");
+
+  int istepStart = 10, iter = 10;
+  elliptic->options.getArgs("CHEBYSHEV START", istepStart);
+  elliptic->options.getArgs("CHEBYSHEV ITER", iter);
+  const int extra = options.compareArgs("CHEBYSHEV EXTRA", "TRUE");
+
+  if (tstep!=tstep_prev) {
+    isolver = 0;
+  }
+  if (isolver>MAXSOLVER) {
+    nrsAbort(platform->comm.mpiComm, EXIT_FAILURE,
+             "%s%d\n", "Chebyshev solver: not enough space sor all solvers!",isolver);
+  }
+  if (istepStart<=0) {
+    nrsAbort(platform->comm.mpiComm, EXIT_FAILURE,
+             "%s%d\n", "Chebyshev solver: Invalid istepStart!",istepStart); 
+  }
+  if (iter<=0) {
+    nrsAbort(platform->comm.mpiComm, EXIT_FAILURE,
+             "%s%d\n", "Chebyshev solver: Invalid iter!",iter); 
+  }
+
+
+  dfloat dmin = dmin_all[isolver];
+  dfloat dmax = dmax_all[isolver];
+  if (verbose && platform->comm.mpiRank == 0) { 
+    printf("%s %d Chebyshev step=%d, start=%d iter=%d extra=%d  dmin %.6e dmax %.6e \n"
+          ,elliptic->name.c_str(),isolver,tstep,istepStart,iter,extra,dmin,dmax);
+  }
+
+  if (tstep < istepStart) {
+    iter = pcg_eigen(elliptic, o_r, o_x, tol, MAXIT, rdotr, dmin, dmax);
+    if (iter>=3) {
+      dmin_all[isolver] = dmin;
+      dmax_all[isolver] = dmax;
+    }
+    // TODO: estimate iteration number 
+  }
+  else {
+    if (verbose && platform->comm.mpiRank == 0) { 
+      printf("CHEBYSHEV ");
+      printf("%s: initial res norm %.15e WE NEED TO GET TO %e \n", elliptic->name.c_str(), rdotr, tol);
+    }
+    ChebyshevSolver(elliptic, o_r, o_x, iter, dmin, dmax, 0);
+
+    // compute res
+    rdotr = platform->linAlg->weightedNorm2Many(
+        mesh->Nlocal,
+        elliptic->Nfields,
+        elliptic->fieldOffset,
+        elliptic->o_invDegree,
+        o_r,
+        platform->comm.mpiComm
+      ) 
+      * sqrt(elliptic->resNormFactor);
+
+    if (extra && rdotr>tol) {
+      const dfloat logConvRate = (log(rdotr) - log(elliptic->res0Norm)) / ((dfloat) iter);
+      const int extraIteration = ((int)  ( (log(tol) - log(rdotr)) / logConvRate ) ) + 1;
+      if (extraIteration>=1) {
+        ChebyshevSolver(elliptic, o_r, o_x, extraIteration, dmin, dmax, iter);
+        iter = iter + extraIteration;
```

Comment:
That will be extra dot product, but I agree it will make sure it converges for general solver.

For GB, since we have the referenced CFL number and referenced resNorm and divNorm, I know when the solver is not converging to the same solution.